accepted/2020/minimum-os-version/minimum-os-version.md

```diff
@@ -0,0 +1,245 @@
+
+Constants for the platform identifiers could perhaps be exposed on a `PlatformIdentifier` type.
+
+The values declared on this attribute are expected to be accessible by extracting constructor arguments from metadata. However, for readability it’s expected to be applied using subclasses with more specific names. To enforce this, it has an internal constructor and its subclasses are sealed.
+
+The concrete subclasses are `IntroducedAttribute`, `ObsoletedAttribute` and `RemovedAttribute` subclasses, which take of the following form:
+
+```csharp
+public sealed class IntroducedAttribute : PlatformAvailabilityAttribute
+{
+    public IntroducedAttribute(
+    string platform,
+    int majorVersion,
+    int minorVersion = 0,
+    int servicingVersion = 0,
+    int bindingRevision = 0,
+    string message = null
+    ) : base(platform, majorVersion, minorVersion, servicingVersion, bindingRevision, message) {}
+}
+```
+
+These attributes are used as follows:
+
+```csharp
+[Introduced(PlatformIdentifier.macOS, 10, 2)]
+[Deprecated(PlatformIdentifier.macOS, 10, 9)]
+[Removed(PlatformIdentifier.macOS, 10, 12, 1)]
+public class Foo : NSObject { ... }
+```
+
+The purpose of keeping ‘removed’ APIs and annotating them with `RemovedAttribute` is to allow apps and libraries to continue to use these APIs in fallback paths for older OS versions. This is a pattern that exists on Apple platforms.
+
+If and when C# support target typing for enums, we could add convenience overloads that would improve readability:
+
+```csharp
+[Introduced (.macOS, 10, 1)]
+[Obsoleted (.macOS, 10, 8)]
+[Removed (.macOS, 10, 12)]
+```
+
+> NOTE: These attributes would only be needed on reference assemblies. They would not be needed at runtime. They could be removed by the linker, and we could also strip them from the implementation assemblies that we ship ourselves.
+
+# Runtime Checks
+
+> NOTE: See [this issue](https://github.com/dotnet/runtime/issues/33331) for a more active/current discussion of runtime check APIs.
+
+The following method will be added to the `RuntimeInformation` class in the `System.Runtime.InteropServices` namespace:
+
+```csharp
+public static bool CheckOSVersion(
+    int majorVersion, int minorVersion = 0, int servicingVersion = 0
+);
+```
+
+It is used as follows:
+
+```csharp
+if (RuntimeInformation.CheckOSVersion(15)) {
+    // use 15.0+ APIs
```

Comment:
Have we thought through the potential for `MissingMethodException`? In the general case, metadata needs to be available to satisfy the JIT. This is not the case for Xamarin.iOS.

Reply:
I think this is a non-issue, and I'm thinking of challenges from the past. Any platform that supports runtime lightup needs to make sure that it is safe to do so. For example, on Windows, this means that a CSWinRT assembly that targets n Windows version needs to itself expose which APIs are safe to call on n-1. And that's what this spec is describing.